interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+	// Update the ID-index mapping of the moved rule.
+	rdb.ids[movedID] = index
+	delete(rdb.ids, id)
+	return rule, nil
+}
+
+// permissionDBForUserSnapInterfacePermission returns the permission DB for the
+// given user, snap, interface, and permission.
+func (rdb *RuleDB) permissionDBForUserSnapInterfacePermission(user uint32, snap string, iface string, permission string) *permissionDB {
+	userSnaps := rdb.perUser[user]
+	if userSnaps == nil {
+		userSnaps = &userDB{
+			PerSnap: make(map[string]*snapDB),
+		}
+		rdb.perUser[user] = userSnaps
+	}
+	snapInterfaces := userSnaps.PerSnap[snap]
+	if snapInterfaces == nil {
+		snapInterfaces = &snapDB{
+			PerInterface: make(map[string]*interfaceDB),
+		}
+		userSnaps.PerSnap[snap] = snapInterfaces
+	}
+	interfacePerms := snapInterfaces.PerInterface[iface]
+	if interfacePerms == nil {
+		interfacePerms = &interfaceDB{
+			PerPermission: make(map[string]*permissionDB),
+		}
+		snapInterfaces.PerInterface[iface] = interfacePerms
+	}
+	permVariants := interfacePerms.PerPermission[permission]
+	if permVariants == nil {
+		permVariants = &permissionDB{
+			VariantEntries: make(map[string]variantEntry),
+		}
+		interfacePerms.PerPermission[permission] = permVariants
+	}
+	return permVariants
+}
+
+// RuleConflict stores the rendered variant which conflicted with that of
+// another rule, along with the ID of that conflicting rule.
+type RuleConflict struct {
+	Variant       string           `json:"pattern-variant"`
+	ConflictingID prompting.IDType `json:"conflicting-rule-id"`
+}
+
+// addRulePermissionToTree adds all the path pattern variants for the given
+// rule to the map for the given permission.
+//
+// If there are conflicting pattern variant froms other rules, all variants
+// which were previously added during this function call are removed
+// from the path map, and an error is returned along with a list of the
+// conflicting variants and the IDs of the conflicting rules.
+func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) (error, []RuleConflict) {
+	permVariants := rdb.permissionDBForUserSnapInterfacePermission(rule.User, rule.Snap, rule.Interface, permission)
+	conflicts := make([]RuleConflict, 0, rule.Constraints.PathPattern.NumVariants())
+	addVariant := func(index int, variant patterns.PatternVariant) {
+		if conflictingVariantEntry, exists := permVariants.VariantEntries[variant.String()]; exists {
+			conflicts = append(conflicts, RuleConflict{
+				Variant:       variant.String(),
+				ConflictingID: conflictingVariantEntry.RuleID,
+			})
+		} else {
+			permVariants.VariantEntries[variant.String()] = variantEntry{
+				Variant: variant,
+				RuleID:  rule.ID,
+			}
+		}
+	}
+	rule.Constraints.PathPattern.RenderAllVariants(addVariant)
+	if len(conflicts) == 0 {
+		return nil, nil
+	}
+	// There were conflicts, so remove any variants which were added to the tree
+	nextMatchIndex := 0
+	removeVariant := func(index int, variant patterns.PatternVariant) {
```

Comment:
Yes, I like this much better. And path patterns might produce duplicate variants, but it doesn't actually matter, since we're adding the same variant entry for any duplicate variants of the same rule anyway.